compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -1276,6 +1276,31 @@ fn prepare_struct_metadata(
 // Tuples
 //=-----------------------------------------------------------------------------
 
+/// Returns names of captured upvars for closures and generators.
+///
+/// Here are some examples:
+///  - `name__field1__field2` when the upvar is captured by value.
+///  - `_ref__name__field` when the upvar is captured by reference.
+fn closure_saved_names_of_captured_variables(tcx: TyCtxt<'tcx>, def_id: DefId) -> Vec<String> {
+    let body = tcx.optimized_mir(def_id);
+
+    body.var_debug_info
+        .iter()
+        .filter_map(|var| {
+            let is_ref = match var.value {
+                mir::VarDebugInfoContents::Place(place) if place.local == mir::Local::new(1) => {
+                    // The projection is either `[.., Field, Deref]` or `[.., Field]`. It
+                    // implies whether the variable is captured by value or by reference.
+                    matches!(place.projection.last().unwrap(), mir::ProjectionElem::Deref)
+                }
+                _ => return None,
+            };
+            let prefix = if is_ref { "_ref__" } else { "" };
+            Some(prefix.to_owned() + &var.name.as_str())
```

Comment:
Could you stick this in `to_symbol` to avoid allocating two strings? At least I don't see a need for it to be separate.

Reply:
For example, considering the following code segment:
```rust
fn main() {
    let mut a = 1;
    let mut b = || {
        a += 1;             // (Y)
    };
    b();                       // (X)
}
```
At the breakpoint `(X)`, we'll print `b` as `b = t::main::{closure#0} {_ref__a: 0x7fffffffddec}`. But at breakpoint `(Y)`, we can print `a` as `a = 1` (instead of `_ref__a = 1`). It's the reason why `_ref__` should be in a separate function.

Reply:
Makes sense, thanks for explaining.